pro_tes/middleware/middleware_pipeline.py

```diff
@@ -0,0 +1,173 @@
+            middlewares (list, optional): A list of middleware components to
+            initialize the pipeline. Defaults to an empty list if not provided.
+
+        Attributes:
+            middlewares (list): A list that holds the middleware components
+            in the pipeline.
+        """
+        self.middlewares = middlewares or []
+        logger.info(f"middleware: {self.middlewares}")
+
+    def add_middleware(self, middleware) -> None:
+        """Add a middleware to the pipeline."""
+        self.middlewares.append(middleware)
+
+    def process_request(
+            self,
+            request: requests.Request,
+            *args,
+            **kwargs
+    ) -> requests.Request:
+        """
+        Process the incoming request through the middleware pipeline.
+
+        This method iterates through the list of middleware components in the
+        pipeline and calls the `set_request` method of each component to
+        process the incoming request.
+
+        Args:
+            request: The incoming requests.Request object.
+            *args: Additional positional arguments to pass to the middleware.
+            **kwargs: Additional keyword arguments to pass to the middleware.
+
+        Returns:
+            requests.Request object tha is the modified request object after
+            processing by all middleware components.
+
+        Raises:
+            ValueError: If a list of middleware components is provided,and all
+            of them fail to process the request, this exception is raised.
+        """
+        for middleware in self.middlewares:
+            logger.info(
+                f"trying to execute the middleware {middleware}"
+            )
+            if isinstance(middleware, list):
+                inner_success = False
+                for mid_instance in middleware:
+                    try:
+                        request = mid_instance.set_request(
+                            request,
+                            *args,
+                            **kwargs
+                        )
+                        inner_success = True
+                        break
+                    except Exception as exc:  # pylint: disable=W0703
+                        logger.exception(
+                            f"Error occurred in middleware: {exc}"
+                        )
+                if not inner_success:
+                    raise ValueError(
+                        f"List of Failed Middlewares {middleware} "
+                    )
+            else:
+                try:
+                    request = middleware.set_request(
+                        request,
+                        *args,
+                        **kwargs
+                    )
+                except Exception as exc:  # pylint: disable=W0703
+                    logger.exception(
+                        f"Error occurred in middleware: {exc}"
+                    )
+        return request
+
+
+class MiddlewareLoader:
```

Comment:
I don't really see the need for a `MiddlewarePipeline` and a `MiddlewareLoader` class at this point. Both of them have just two public methods and a single attribute that's basically a list of middlewares. Why not just merge them into a single `MiddlewareHandler` class?